tools/embedding_convert.lua

```diff
@@ -0,0 +1,343 @@
+require('onmt.init')
+require('torch')
+local tds = require('tds')
+local zlib = require ('zlib')
+local path = require('pl.path')
+
+local cmd = torch.CmdLine()
+
+
+cmd:text("")
+cmd:text("**embedding_convert.lua**")
+cmd:text("")
+
+cmd:option('-config', '', [[Read options from this file]])
+
+cmd:text("")
+cmd:text("**Data options**")
+cmd:text("")
+
+cmd:option('-dict_file', '', [[Path to outputted dict file from preprocess.lua.]])
+cmd:option('-embed_file', '',[[Path to embedding file. Ignored if auto_lang is used.]])
+cmd:option('-save_data', '',[[Output file path/label]])
+
+cmd:text("")
+cmd:text("**Embedding options**")
+cmd:text("")
+
+cmd:option('-auto_lang', '', [[Wikipedia Language Code to autoload embeddings.]])
+cmd:option('-embed_type', 'word2vec',[['word2vec' or 'glove'. Ignored if auto_lang is used.]])
+cmd:option('-normalize', 'true',[[Boolean to normalize the word vectors, or not.]])
+cmd:option('-report_every', '100000',[[Print stats every this many lines read from embedding file.]])
+
+local opt = cmd:parse(arg)
+
+
+
+-- [[Auto Loads language files from S3]]
+-- [[Embedding files made available by Rami Al-Rfou through Polygot ( Project: https://pypi.python.org/pypi/polyglot Paper: http://www.aclweb.org/anthology/W13-3520 )]]
+local function loadAuto(lang)
+
+  local http = assert(require("socket.http"), 'autoload requires \'luasocket\' which can be installed via \'luarocks install luasocket\'')
+
+  --TODO: Perhaps centralize?
+  local endpoint = 'http://language-embedding-files.s3-website-us-east-1.amazonaws.com/'
+
+
+  local filename = 'polyglot-' .. lang:lower() .. '.txt'
+  local filepath = opt.save_data .. "-" .. filename
+
+  if path.exists(filepath) then
+    return filepath
+  end
+
+  local resp, stat = http.request(endpoint .. filename .. '.gz')
+
+  if stat == 200 then
+
+    local autoFile = io.open(filepath, "w")
+
+    local result = zlib.inflate()(resp)
+
+    autoFile:write(result)
+    autoFile:close()
+
+  else
+
+    error('embedding file for language code \'' .. lang .. '\' was not found')
+
+  end
+
+  return filepath
+
+end
+
+
+
+local function loadEmbeddings(embeddingFilename, embeddingType, dictionary)
+
+
+  --[[Converts binary to strings - Courtesy of https://github.com/rotmanmi/word2vec.torch]]
+  local function readStringv2(file)
+    local str = {}
+    local max_w = 50
+
+    for _ = 1, max_w do
+      local char = file:readChar()
+
+      if char == 32 or char == 10 or char == 0 then
+        break
+      else
+        str[#str + 1] = char
+      end
+    end
+
+    str = torch.CharStorage(str)
+    return str:string()
+
+  end
+
+  -- [[Looks for cased version and then lower version of matching dictionary word.]]
+  local function locateIdx(word, dict)
+
+    local idx = nil
+
+    if dict:lookup(word) ~= nil then
+      idx = dict:lookup(word)
+
+    elseif dict:lookup(word:lower()) ~= nil then

```

Comment:
Since lookup returns nil, you could optimize and do the lookup only once.
Maybe something like:
`    idx = dict:lookup(word) 
    if idx == nil then
      idx = dict:lookup(word:lower())
    end`